### PR TITLE
enforce max size on firehose columns in ruby

### PR DESCRIPTION
**Pull request by davidsbailey**
Suresh helpfully deduced that firehose imports sometimes fail to make it into redshift when some of the columns are too large (see [slack](https://codedotorg.slack.com/archives/C0T0PNTM3/p1705518125559239)).

some redshift querying suggests the max size of the data_json column that has made it from firehose into redshift without failing the import step is 65526 bytes:
Screenshot 2024-01-18 at 2 08 39 PM

this PR makes the limit explicit, so that we get a honey badger notification for the single offending firehose record, rather than a silent failure affecting all records in the batch.

## Testing story

new unit tests

### CODE REVIEW

**Comment by sureshc** [comment]
Can we also update our browser Firehose client with the same data validation rule (and publish an error to NewRelic when the Firehose event can't be published)? https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/lib/util/firehose.js

**Comment by sureshc** [comment]
Querying the schema confirms that the width of the `data_json` column is 65535. Looking at one of the sample failed events that didn't load, it's possible we're exceedingly the `data_string` 4096 limit in some cases, so we should probably validate that field as well. I haven't inspected our Firebase code to be certain, yet.

``` SQL
SELECT 
    column_name, 
    data_type, 
    character_maximum_length,
    numeric_precision,
    numeric_scale,
    is_nullable
FROM 
    information_schema.columns 
WHERE 
    table_schema = 'analysis' 
    AND table_name = 'events'
ORDER BY 
    ordinal_position;
```

image

**Comment by davidsbailey** [comment]
good catch! I see several events in study `regional-partner-search-log` with size close to 4096, so that matches up with the idea that some of them may exceed that limit.

**Comment by davidsbailey** [comment]
> Can we also update our browser Firehose client with the same data validation rule (and publish an error to NewRelic when the Firehose event can't be published)? https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/lib/util/firehose.js

another good catch. I would love some help with this. @hannahbergam would you be willing to pick up the work Suresh is describing on the js side? this PR should serve as a template for the work you need to do there, and once it's done we believe that firehose will be reliable again.

@sureshc, regardless, I'd love to have your review on this PR for the ruby side, and have whatever implementation we do in js come in a later PR.

File: lib/cdo/firehose.rb [hunk 1]
```diff
@@ -123,6 +129,12 @@ def put_record(stream, data)
...
 
+    if data[:data_json].to_s.bytesize > MAX_DATA_JSON_BYTES
+      raise ArgumentError.new("data_json column too large (#{data[:data_json].to_s.bytesize} bytes)")
+    end
+    if data[:data_string].to_s.bytesize > MAX_DATA_STRING_BYTES
+      raise ArgumentError.new("data_string column too large (#{data[:data_string].to_s.bytesize} bytes)")
+    end
     record = add_common_values(data).to_json
     if (size = record.bytesize) > BYTES_PER_RECORD
       raise ArgumentError.new("Record too large (#{size} bytes)")
```
Comment (hannahbergam) [inline lib/cdo/firehose.rb:140]: Thanks for adding the new error catches Dave! Can you help me understand how data that that exceeds the max exits this function and still make it into firehose? I've read through the file but can't tell from a first read.

Comment (davidsbailey) [reply lib/cdo/firehose.rb:140]: data that exceeds the max does exit this function, but will not still make it into firehose. does that help?

Comment (sureshc) [reply lib/cdo/firehose.rb:140]: It’s pretty convoluted and it was mostly configured manually in AWS, so there isn’t much code that you can search for showing how it works. Here’s my memory of the architecture:

1. Various components in our Ruby web application servers invoke this library (`firehouse.rb`) to publish an Event 
2. This library publishes an Event in an AWS Firehose Stream
3. The AWS Firehose stream writes events as JSON text in batch text files in AWS S3 Objects (each file contains a long list of JSON objects).
4. A different AWS Firehose stream asynchronously picks ups those S3 Objects and executes a SQL statement on our Redshift cluster to INSERT each of the JSON objects into the Redshift `analysis.events` table, mapping specific properties in the JSON object to columns in the `events` table.
5. If a single JSON event within a batch file contains a value in one of its attributes that violates the data integrity rules (length, datatype, etc.) of the corresponding column in the `analysis.events` table THE ENTIRE BATCH OF EVENTS fails to load into Redshift. This means that a single event that has a `data_json` or `data_string` attribute that’s too long, can cause the other 1-2 thousand events in that batch to fail to get delivered to Reshift :(

By validating the data BEFORE publishing the Firehouse Event, this Pull Request will dramatically improve the reliability of end-to-end delivery of events.

File: lib/cdo/firehose.rb [hunk 2]
```diff
@@ -57,6 +57,12 @@ class FirehoseClient
   # Ref: https://docs.aws.amazon.com/firehose/latest/dev/limits.html
   BYTES_PER_REQUEST = 1024 * 1024 * 4
 
+  # This limit is based on the empirical observation that no data_json column values larger than 65,526 bytes have
```
Comment (sureshc) [inline lib/cdo/firehose.rb:60]: We could note the exact length of the destination Redshift database columns where these 2 attributes get delivered.

**Review by sureshc — APPROVED** [review]
We (Infrastructure) have had this data validation on our backlog for a couple of years. Thank you for getting this done for us!

File: lib/cdo/firehose.rb (hunk 1 above)
Comment (hannahbergam) [reply lib/cdo/firehose.rb:140]: Okay this is helpful- thank you both! I now realize that we are sacrificing a small amount of data so it doesn't fail the entire batch. So firehose will be *slightly undercounting events, but on a tiny scale?

Comment (davidsbailey) [reply lib/cdo/firehose.rb:140]: > So firehose will be *slightly undercounting events, but on a tiny scale?

yes, but only for firehose events which break the rules, and we'll now get notified about these cases so we can fix them. we should assume that the vast majority of firehose event types are written in a way that it'll never break these rules.

**Pull request merged**